### ISSUE TITLE
ECMAScriptExpressions: Support JSON.stringify(inputs)

### ISSUE THREAD

**Issue by cjllanwarne** [broadinstitute/cromwell#3090]
EDIT: to safe yourself some data entry, you can use branch [cjl_initial_work_dir_requirement_4](https://github.com/broadinstitute/cromwell/tree/cjl_initial_work_dir_requirement_4) as an entry point with the centaur test and a Spec already added.

This seems to be a pretty common pattern but relies on `JSON.stringify(inputs)` working in our expression evaluator:
```yml
# A common use case: stringy the inputs JSON and provide that file as another input file.

cwlVersion: v1.0
$graph:
- id: stringify_inputs
  class: CommandLineTool
  baseCommand: ['grep', 'number', 'inputs.json']
  requirements:
      - class: DockerRequirement
        dockerPull: "python:3.5.0"
      - class: InitialWorkDirRequirement
        listing:
            - entryname: 'inputs.json'
              entry: $(JSON.stringify(inputs))

  stdout: "number_field"

  # TODO CWL: Set the types more appropriately (depends on issue #3059)
  inputs:
      - id: number
        type: string
        default: 27
      - id: str
        type: string
        default: wooooo
      - id: boolean
        type: string
        default: True
  outputs:
      - id: number_field_output
        type: string
        outputBinding:
          glob: number_field
          loadContents: true
          outputEval: $(self[0].contents.trim())
```

**Comment by kshakir** [comment]
**TL;DR Good news: I can fix JSON.stringify in my upcoming PR. Bad news: it breaks other JSON**

```java
import jdk.nashorn.api.scripting.AbstractJSObject;
import jdk.nashorn.api.scripting.ClassFilter;
import jdk.nashorn.api.scripting.JSObject;
import jdk.nashorn.api.scripting.NashornScriptEngineFactory;

import javax.script.Bindings;
import javax.script.ScriptContext;
import javax.script.ScriptEngine;
import javax.script.ScriptException;
import javax.script.SimpleScriptContext;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.Set;

/**
 * Example showing that a JSObject will not cast a property key to string, plus an example showing how JSON.stringify
 * returns null for Java maps. So:
 * - If you need non-string key expressions ex: 'myObj[true]', then do not pass in a JSObject.
 * - If you need to use JSON.stringify, then do not pass in a java.util.Map.
 */
public class Main {
    public static void main(String[] args) throws ScriptException {
        testKeyCastImplicit();
        testKeyCastExplicit();
        System.out.println("---");
        testKeyMapString();
        testKeyMapNonString();
        System.out.println("---");
        testStringifyMap();
        testStringifyJSObject();
        testStringifyScriptObject();
    }

    /**
     * Create and use an object in Javascript, return it to Java, and then pass it back into similar Javascript. In the
     * second Javascript call we cannot use our key in the same way.
     */
    private static void testKeyCastImplicit() throws ScriptException {
        final String expr1 = expr(
                "var obj = {};",
                "obj[true] = 'world';",
                "print('testKeyCastImplicit hello init ' + obj[true]);",
                "obj;"
        );
        final Map<String, Object> args1 = Collections.emptyMap();
        final JSObject result1 = (JSObject) eval(expr1, args1);
        
        /*
        JSObjectLinker does not cast keys to strings during get (and similarly in put).
        - https://github.com/JetBrains/jdk8u_nashorn/blob/jdk8u76-b03/src/jdk/nashorn/internal/runtime/linker/JSObjectLinker.java#L179
        - http://hg.openjdk.java.net/jdk9/jdk9/nashorn/file/17cc754c8936/src/jdk.scripting.nashorn/share/classes/jdk/nashorn/internal/runtime/linker/JSObjectLinker.java#l190

        According to ECMA, property accessors in brackets are expressions and not identifier-name-string:

        > 3. Let propertyNameReference be the result of evaluating Expression.
        > 4. Let propertyNameValue be GetValue(propertyNameReference).
        > ...
        > 6. Let propertyNameString be ToString(propertyNameValue).

        - https://www.ecma-international.org/ecma-262/5.1/#sec-11.2.1

        Similarly according to the MDN:

        > Property names must be strings. This means that non-string objects cannot be used as keys in the object. Any
        > non-string object, including a number, is typecasted into a string via the toString method.

        - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Property_Accessors#Property_names

        This is not totally unexpected as according to the Nashorn engine notes:

        > not every operation and script API (JSON, Array, Function's properties/functions) treats ScriptObjectMirror
        > and jdk.nashorn.internal.runtime.ScriptObject uniformly. There are places where ScriptObjects work as
        > expected but if you pass ScriptObjectMirror or your own JSObject implementation, it won't work as expected.

        - https://wiki.openjdk.java.net/display/Nashorn/Nashorn+jsr223+engine+notes

        See also:

        - https://stackoverflow.com/questions/44232013/is-there-a-way-to-make-a-custom-implementation-of-nashorn-jsobject-work-with-obj#44246540
         */
        final String expr2 = expr("print('testKeyCastImplicit hello again ' + obj[true]);");
        final Map<String, Object> args2 = Collections.singletonMap("obj", result1);
        eval(expr2, args2);
    }

    /**
     * Create and use an object in Javascript, return it to Java, and then pass it back into similar Javascript. In the
     * second Javascript call we only get a non-null result when the key is explicitly cast to a string.
     */
    private static void testKeyCastExplicit() throws ScriptException {
        final String expr1 = expr(
                "var obj = {};",
                "obj[true] = 'world';",
                "print('testKeyCastExplicit hello init ' + obj[true]);",
                "obj;"
        );
        final Map<String, Object> args1 = Collections.emptyMap();
        final JSObject result1 = (JSObject) eval(expr1, args1);

        final String expr2 = expr("print('testKeyCastExplicit hello again ' + obj[true.toString()]);");
        final Map<String, Object> args2 = Collections.singletonMap("obj", result1);
        eval(expr2, args2);
    }

    /**
     * Show that keys do not coerce to a string in a java.util.Map.
     */
    private static void testKeyMapString() throws ScriptException {
        /*
        See notes in testKeyCastImplicit regarding keys should be cast to strings.

        Because ECMA says that keys are strings, this lookup should work but doesn't.
         */
        final String expr = expr(
                "print('testKeyMapString hello ' + obj[true]);",
                "obj;"
        );
        final Map<String, Object> obj = Collections.singletonMap("true", "world");
        final Map<String, Object> args = Collections.singletonMap("obj", obj);
        eval(expr, args);
    }

    /**
     * Show that key equality works for non-strings in a java.util.Map.
     */
    private static void testKeyMapNonString() throws ScriptException {
        /*
        See notes in testKeyCastImplicit regarding keys should be cast to strings.

        Keys are not coerced to strings upon lookup nor internally, meaning we actually get a match on a Map with
        non-string keys.

        Ideally instead of passing the lookup directly to java.util.Map.get() a coerced string should be passed. However
        if the lookup keys are going to be coerced, the map's keys should have already been coerced also. The internal
        map must contain string keys anyway to be ECMA compliant. Thus a java.util.Map<Object, Object> should be
        converted to a Map<String, Object> where the keys are coerced to strings. Ideally this test should still work
        after any fix OR Nashorn should not explicitly not support java.util.Map with non-string keys.

        For now, the JSObjectLinker is directly invoking the map getter:

        - https://github.com/JetBrains/jdk8u_nashorn/blob/jdk8u76-b03/src/jdk/nashorn/internal/runtime/linker/JSObjectLinker.java#L84-L86
        - http://hg.openjdk.java.net/jdk9/jdk9/nashorn/file/17cc754c8936/src/jdk.scripting.nashorn/share/classes/jdk/nashorn/internal/runtime/linker/JSObjectLinker.java#l87

        The BeanLinker generates the dynamic call to the map getter:

        - https://github.com/JetBrains/jdk8u_nashorn/blob/jdk8u76-b03/src/jdk/internal/dynalink/beans/BeanLinker.java#L179-L205
        - http://hg.openjdk.java.net/jdk9/jdk9/nashorn/file/17cc754c8936/src/jdk.dynalink/share/classes/jdk/dynalink/beans/BeanLinker.java#l218
         */
        final String expr = expr(
                "print('testKeyMapNonString hello ' + obj[true]);",
                "obj;"
        );
        final Map<Object, Object> obj = Collections.singletonMap(true, "world");
        final Map<String, Object> args = Collections.singletonMap("obj", obj);
        eval(expr, args);
    }

    /**
     * Fail to stringify a java.util.Map.
     */
    private static void testStringifyMap() throws ScriptException {
        /*
        Nashorn's JSON.stringify identifies that a map is an object, but doesn't convert Java maps (nor Java arrays) to
        a string. NOTE: There has been some work in jdk8u and jdk9 to fix issues with stringify, so the code varies a
        bit depending on the jdk version.

        - https://github.com/JetBrains/jdk8u_nashorn/blob/jdk8u76-b03/src/jdk/nashorn/internal/objects/NativeJSON.java#L267-L272
        - http://hg.openjdk.java.net/jdk9/jdk9/nashorn/file/17cc754c8936/src/jdk.scripting.nashorn/share/classes/jdk/nashorn/internal/objects/NativeJSON.java#l288

        Maps do get special treatment in Nashorn and are passed around internally:

        - https://wiki.openjdk.java.net/display/Nashorn/Nashorn+extensions#Nashornextensions-SpecialtreatmentofobjectsofspecificJavaclasses

        In this case the value within NativeJSON.str() is the Java map instance and not a ScriptObject nor a JSObject.
        Thus the code falls through and returns UNDEFINED. If the code didn't bypass the if-statement the block would
        run NativeJSON.JO() and attempt to get the value's keys.

        - https://github.com/JetBrains/jdk8u_nashorn/blob/jdk8u76-b03/src/jdk/nashorn/internal/objects/NativeJSON.java#L287
        - http://hg.openjdk.java.net/jdk9/jdk9/nashorn/file/17cc754c8936/src/jdk.scripting.nashorn/share/classes/jdk/nashorn/internal/objects/NativeJSON.java#l311

        Even in the latest code, getting an object's keys is not expecting a java.util.Map:

        - http://hg.openjdk.java.net/jdk9/jdk9/nashorn/file/17cc754c8936/src/jdk.scripting.nashorn/share/classes/jdk/nashorn/internal/objects/NativeJSON.java#l443

        There is also bug 8181203 that mentions that Object.keys should work on any JSObject, not just
        ScriptObjectMirror.

        - https://bugs.openjdk.java.net/browse/JDK-8181203

        As maps have special treatment in Nashorn, perhaps stringify should be expecting these Java maps (and Java
        arrays) also?

        See also:

        - https://www.ecma-international.org/ecma-262/5.1/#sec-15.12.3
        - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify
         */
        final String expr = expr(
                "print('testStringifyMap type ' + typeof(obj));",
                "print('testStringifyMap json ' + JSON.stringify(obj));"
        );
        final Map<String, Object> obj = Collections.singletonMap("true", "world");
        final Map<String, Object> args = Collections.singletonMap("obj", obj);
        eval(expr, args);
    }

    /**
     * Stringify a JSObject.
     */
    private static void testStringifyJSObject() throws ScriptException {
        // JSON.stringify works fine on JSObject
        final String expr = expr(
                "print('testStringifyJSObject type ' + typeof(obj));",
                "print('testStringifyJSObject json ' + JSON.stringify(obj));"
        );
        final JSObject obj = new AbstractJSObject() {
            // @formatter:off
            final Map<String, Object> map = Collections.singletonMap("true", "world");
            @Override public Object getMember(String name) { return map.get(name); }
            @Override public boolean hasMember(String name) { return map.containsKey(name); }
            @Override public Set<String> keySet() { return map.keySet(); }
            @Override public Collection<Object> values() { return map.values(); }
            // @formatter:on
        };
        final Map<String, Object> args = Collections.singletonMap("obj", obj);
        eval(expr, args);
    }

    /**
     * Stringify a ScriptObject.
     */
    private static void testStringifyScriptObject() throws ScriptException {
        // JSON.stringify works fine on ScriptObject
        final String expr = expr(
                "var obj = {};",
                "obj[true] = 'world';",
                "print('testStringifyScriptObject type ' + typeof(obj));",
                "print('testStringifyScriptObject json ' + JSON.stringify(obj));"
        );
        final Map<String, Object> args = Collections.emptyMap();
        eval(expr, args);
    }

    private static Object eval(final String expr, final Map<String, Object> values) throws ScriptException {
        // The engine could be static, but for example purposes making a new engine/context/bindings per evaluation.
        final ScriptEngine engine =
                ENGINE_FACTORY.getScriptEngine(nashornStrictArgs, getNashornClassLoader(), noJavaClassFilter);
        final Bindings bindings = engine.createBindings();
        bindings.putAll(values);

        final ScriptContext context = new SimpleScriptContext();
        context.setBindings(bindings, ScriptContext.ENGINE_SCOPE);
        return engine.eval(expr, context);
    }

    private static final NashornScriptEngineFactory ENGINE_FACTORY = new NashornScriptEngineFactory();

    /**
     * Add stricter Nashorn arguments to the default `-doe`.
     *
     * @see <a href="https://docs.oracle.com/javase/8/docs/technotes/tools/windows/jjs.html">JJS docs and options</a>
     * @see <a href="https://github.com/JetBrains/jdk8u_nashorn/blob/jdk8u76-b03/src/jdk/nashorn/internal/runtime/resources/Options.properties">Nashorn supported options (github) </a>
     * @see <a href="http://hg.openjdk.java.net/jdk8/jdk8/nashorn/file/5dbdae28a6f3/src/jdk/nashorn/internal/runtime/resources/Options.properties">Nashorn supported options</a>
     * @see jdk.nashorn.api.scripting.NashornScriptEngineFactory#DEFAULT_OPTIONS
     */
    private static String[] nashornStrictArgs = {
            "-doe", "-strict", "--no-java", "--no-syntax-extensions", "--language=es5"
    };

    /**
     * Don't allow any java classes.
     */
    private static ClassFilter noJavaClassFilter = anyClass -> false;

    /**
     * Copy/paste of the private jdk.nashorn.api.scripting.NashornScriptEngineFactory#getAppClassLoader().
     * Of all the overloads of jdk.nashorn.api.scripting.NashornScriptEngineFactory#getScriptEngine, there is no jdk8
     * one that receives just args and a class filter.
     *
     * @see jdk.nashorn.api.scripting.NashornScriptEngineFactory#getScriptEngine
     * @see jdk.nashorn.api.scripting.NashornScriptEngineFactory#getAppClassLoader
     */
    private static ClassLoader getNashornClassLoader() {
        // Revisit: script engine implementation needs the capability to
        // find the class loader of the context in which the script engine
        // is running so that classes will be found and loaded properly
        final ClassLoader ccl = Thread.currentThread().getContextClassLoader();
        return (ccl == null) ? NashornScriptEngineFactory.class.getClassLoader() : ccl;
    }

    private static final String LINE_SEPARATOR = System.getProperty("line.separator");

    private static String expr(String... lines) {
        return String.join(LINE_SEPARATOR, lines);
    }
}
```
